### DISCUSSION TITLE
Improve Ecotone performance

### DISCUSSION THREAD

**Discussion by jlabedo** [ecotoneframework/ecotone-dev#222] [General]
So let me present my thoughts on performance improvement.

Here is a screenshot of an xdebug profile of ecotone lite "PlaceOrder" benchmark in production, with 100 iterations to avoid autoloader impact, and opcache disabled. The bootstraping of the messaging system is 63%, the execution of the test case is 36% of the time. Fail fast is disabled, so some messaging components are also built during execution.

![image](https://github.com/ecotoneframework/ecotone-dev/assets/1191198/3f2dccf9-25b2-461d-b222-1e04353d16ad)

# I see mainly three ways of improving performance

## Messaging system resolution
I will take for example the gateway proxy builder that is one of the most time consuming component in the profile (36%).
Looking at the code, a lot of resolutions are taken during build: resolving gateway converters, resolve annotations, sort interceptors, add an interceptor for error channel. A lot of this could be resolved during preparation phase and be cached.

## Opcache and messaging system serialization
In one of my previous tests, I tried to export the messaging system configuration as a php file instead of serializing it. The performance gain was more or less 50% with opcache enabled, which should be the case for http requests.

## Lazy component creation
Some components are not lazy: for instance all gateways are created on each boot.

**Comment by jlabedo** [comment]
## Solution 1: resolve all messaging components during preparation phase for caching
During preparation phase, builders (`GatewayProxyBuilder` for example) could be compiled to resolve all they can, like interceptors, parameter converters, gateway converters, annotations, ... We could add this interface that a builder could implement. :
```php
interface CompilableBuilder {
  public function compile(ReferenceSearchService $referenceSearchService, ChannelResolver $channelResolver): void;
}
```
After "compilation", builders should be locked for changes and immutable.
During the execution phase, a typical implementation of a builder would be:
```php
class GatewayProxyBuilder implements InterceptedEndpoint, CompilableBuilder
{
//...
  public function build(ReferenceSearchService $referenceSearchService, ChannelResolver $channelResolver) {
    $interfaceToCall = $interfaceToCallRegistry->getFor($this->interfaceName, $this->methodName);
    if ($this->isCompiled) {
      return new Gateway(
        $interfaceToCall,
        new MethodCallToMessageConverter(
            $interfaceToCall,
            $this->methodArgumentConverters
        ),
        $this->messageConverters,
        new GatewayReplyConverter(
            $referenceSearchService->get(ConversionService::REFERENCE_NAME),
            $interfaceToCall,
            $this->messageConverters,
        ),
        $this->buildGatewayInternalHandler($interfaceToCall, $referenceSearchService, $channelResolver)
      );
    } else {
      // ...
    }
  }
}
```

This solution would help only on "Messaging system resolution" performance.

**Comment by jlabedo** [comment]
# Solution 2: dump a messaging system into a container

Continuing on solution 1, we could think of the `MessagingSystemConfiguration` like a ContainerBuilder and the `MessagingSystem` like a Container of messaging components. The Messaging configuration and all components builder could return `definitions` of messaging system components, like in php-di or symfony. Here is an example:
```php
class GatewayProxyBuilder implements InterceptedEndpoint, CompilableBuilder
{
//...
  public function build(ReferenceSearchService $referenceSearchService, ChannelResolver $channelResolver) {
    $interfaceToCallReference = new InterfaceToCallReference($this->interfaceName, $this->methodName)
    // All resolutions can be done during this phase
    // And then dumped into "definitions"
    return new Definition(Gateway::class, [
      $interfaceToCallReference,
      new Definition(MethodCallToMessageConverter::class, [
          $interfaceToCallReference,
          array_map($this->methodArgumentConverters, ($converter) => $converter->build())
      ]),
      ...
    );
  }
}
```
With these definitions, we would be able to dump all messaging system components into an highly performant php container. With this solution, we would have:
- improved messaging system resolutions. Everything would be resolved during preparation phase.
- opcache optimization: container implementation can dump a container to a php file, so opcache can optimize the code.
- lazy creation: container implementations have lazy creation built in

A proof of concept is #223.

**Comment by dgafka** [comment]
@jlabedo I would like to see the profiler results after latest changes on `1.107.1`. This does fixes the problem with to early constructed gateways and optimize the way we create proxies. This solution still needs some optimalization phase, as right now building checks are ran on first execution.

I've not yet invested much time into the POC, but I understood the idea. I wonder about what will be performance gains now with the above optimalizations. Is this something you can verify early or we need to wait for POC to be finished to see the benchmarks?

**Reply by jlabedo** [reply]
It helps :
<img width="1177" alt="Capture d’écran 2023-10-07 à 10 24 24" src="https://github.com/ecotoneframework/ecotone-dev/assets/1191198/9f4185d2-4762-4015-be77-b3177edbd4ee">

A container implementation will help lazy load every component (and checking consistency during compiling phase).

Unfortunately, my PoC is not showing any substantial performance improvement (and only thanks to lazy loading gateways as you did in your last PR). I am pretty sure the performance boost would be like x2 to x3 on PlaceOrderBenchmark with opcache when PoC is complete, but I have to go deeper: I think I have to complete the refactoring with at least Core and async modules to run PlaceOrderBenchmark to finish the PoC.

**Reply by jlabedo** [reply]
@dgafka , I managed to get a PoC working (it works only for the PlaceOrderBenchmark simple case). I get a 3,8x performance gain on this benchmark ! The tests are not working as I refactored only the components necessary for this benchmark to run without error.

**Reply by dgafka** [reply]
wow that sound amazing @jlabedo :) 

Let me know when you will get PoC ready in a shape you would like me to review :)

**Reply by jlabedo** [reply]
I would like to validate some choices before going further, do you mind if I spot and comment intersting parts in the code @dgafka ?
The problem is that I thought we could go step by step, but it adds complexity and no performance improvement. So my favorite choice would be to move every components in one shot. It's not complex but It's time consuming and hard to rebase 😅

**Reply by dgafka** [reply]
@jlabedo the most important is to agree on the final solution. If this PR will at the end get to complex we can split, but with agreement we we will know what is the aim. 

Please do comment the parts, you would like to discuss :)

**Reply by dgafka** [reply]
I did partial review. 
I like that the Container Definition abstraction is hold within the framework. 
What I was bit worried about, is that we will now need to integrate with external compilation process for mission critical components. As large integration surface creates time consuming maintenance. 
But it looks to me now, that this can actually simplify the process of the integration e.g. dropping `resolved references` and framework specific container implementation surface is rather small.

**Reply by jlabedo** [reply]
I understand your concern about integrating with external integration process (and I knew you would have this concern :) ). As you can see, it is relatively small and dumb to integrate with Symfony or php-di. For Laravel I don't know (help would be very welcomed).

I agree that depending on different frameworks for mission critical components seems dangerous, as it can result in "it works on Symfony but not in Laravel", or "it works with Symfony 6 but not with Symfony 7". But:
- there is less code to maintain
- framework container implementations are mature and stable
- the responsibility of the framework is small, it's only instantiation.

Moreover, if we decide to return to a single container implementation for all frameworks for some reason, it won't change a lot.

**Reply by dgafka** [reply]
Yea, I can see it's small integration surface. 
So let's proceed with the idea of framework specific implementations of the containers. 
About Laravel, I believe it use Symfony DI under the hood. We can verify that at the end of implementation and right now focus on Ecotone Lite and Symfony.

**Reply by jlabedo** [reply]
Great, thanks for your review, you spotted the interesting parts.

**Reply by jlabedo** [reply]
@dgafka , do you agree with the one big refactoring without progressive refactoring ?

**Reply by dgafka** [reply]
Yea I agree @jlabedo. I can see it will be easier to prepare everything at once and just fix the code along the way, than try to make a split now (not even sure what splits we would apply here). Better to put energy into finishing and testing this up.
We've high test coverage for the business functionality, so if we manage to pass tests, we've high assurance that it simply works.
